library/objective-c/EnvoyDeviceLifecycleMonitor.m

```diff
@@ -0,0 +1,35 @@
+#import "library/objective-c/EnvoyEngine.h"
+
+#import "library/common/main_interface.h"
+
+#import <UIKit/UIKit.h>
+
+@implementation EnvoyDeviceLifecycleMonitor
+
++ (void)startFlushingStatsOnLifecycleChanges {
+  static dispatch_once_t lifecycleMonitoringStarted;
+  dispatch_once(&lifecycleMonitoringStarted, ^{
+    [self startObservingLifecycleNotifications];
+  });
+}
+
+#pragma mark - Private
+
++ (void)startObservingLifecycleNotifications {
+  NSNotificationCenter *notificationCenter = [NSNotificationCenter defaultCenter];
+  [notificationCenter addObserver:self
+                         selector:@selector(lifecycleDidChangeWithNotification:)
+                             name:UIApplicationWillResignActiveNotification
+                           object:nil];
+  [notificationCenter addObserver:self
+                         selector:@selector(lifecycleDidChangeWithNotification:)
+                             name:UIApplicationWillTerminateNotification
+                           object:nil];
+}
+
++ (void)lifecycleDidChangeWithNotification:(NSNotification *)notification {
+  NSLog(@"[Envoy] triggering stats flush (%@)", notification.name);
```

Comment:
Is this meant to be here? (Do we generally utilize NSLog in our platform layer? Is it normal for third-part libraries to do so?)

Reply:
We have several of these at the moment. @junr03 and I briefly discussed the possibility of using `ENVOY_LOG` rather than `NSLog` (which would be much nicer), but it'd require the native layer to depend on upstream Envoy code directly which would be a bit of a shift. Opened an issue for this here: https://github.com/lyft/envoy-mobile/issues/719

Reply:
Sounds good, and thanks for opening the issue.